lib/src/sql/parser.rs

```diff
@@ -159,6 +229,79 @@ mod tests {
 		assert!(res.is_err());
 	}
 
+	#[test]
+	fn parse_ok_recursion() {
+		let sql = "SELECT * FROM ((SELECT * FROM (5))) * 5;";
+		let res = parse(sql);
+		assert!(res.is_ok());
+	}
+
+	#[test]
+	fn parse_ok_recursion_deeper() {
+		let sql = "SELECT * FROM (((( SELECT * FROM ((5)) + ((5)) + ((5)) )))) * ((( function() {return 5;} )));";
+		let start = Instant::now();
+		let res = parse(sql);
+		let elapsed = start.elapsed();
+		assert!(res.is_ok());
+		assert!(elapsed < Duration::from_millis(150), "previously took ~15ms in debug")
+	}
+
+	#[test]
+	fn parse_recursion_cast() {
+		for p in 1..=4 {
+			recursive("SELECT * FROM ", "<int>", "5", "", 10usize.pow(p), p > 1);
+		}
+	}
+
+	#[test]
+	fn parse_recursion_geometry() {
+		for n in [3, 40, 100] {
+			recursive(
+				"SELECT * FROM ",
+				r#"{type: "GeometryCollection",geometries: ["#,
+				r#"{type: "MultiPoint",coordinates: [[10.0, 11.2],[10.5, 11.9]]}"#,
+				"]}",
+				n,
+				n > 30,
+			);
+		}
+	}
+
+	#[test]
+	fn parse_recursion_javascript() {
+		for p in 1..=3 {
+			recursive("SELECT * FROM ", "function() {", "return 5;", "}", 10usize.pow(p), p > 1);
+		}
+	}
+
+	#[test]
+	fn parse_recursion_mixed() {
+		for n in [3, 8] {
+			recursive("", "SELECT * FROM ((((", "5 * 5", ")))) * 5", n, n > 3);
+		}
+	}
+
+	#[test]
+	fn parse_recursion_select() {
+		for p in 1..=3 {
+			recursive("SELECT * FROM ", "(SELECT * FROM ", "5", ")", 6usize.pow(p), p > 1);
+		}
+	}
+
+	#[test]
+	fn parse_recursion_value_subquery() {
+		for p in 1..=4 {
+			recursive("SELECT * FROM ", "(", "5", ")", 10usize.pow(p), p > 1);
+		}
+	}
+
+	#[test]
+	fn parse_recursion_if_subquery() {
+		for p in 1..=3 {
+			recursive("SELECT * FROM ", "IF true THEN ", "5", " ELSE 4 END", 6usize.pow(p), p > 1);
+		}
+	}
+
```

Comment:
Based on the comment above, I would also test:
* dive until the MAX_DEPTH
* drop the last parser
* verify you can dive again

Not sure how complex will it be to implement this test case

Reply:
Turns out this error caused the majority of tests (and some benchmarks) to fail. Rather than adding a dedicated test, I have centralized and documented the value of `DEPTH_PER_DIVE` in a constant.

Reply:
sounds good 👌